Add Tic Tac Toe game

janaSunrise opened the pull request.

janaSunrise committed:
Add some more fun commands

ItsDrike commented on cogs/fun.py:
```diff
@@ -224,7 +292,7 @@ async def yesno(self, ctx: Context, *, question: str) -> None:
                     await ctx.send(f"OMFG! [STATUS : {meme.status}]")
 
     @command(aliases=["dadjoke", "awdad", "dadpls", "shitjoke", "badjoke"])
-    async def joke(self, ctx: Context) -> None:
+    async def dadjoke(self, ctx: Context) -> None:
```
Since command name is changed to `dadjoke` there is no point in keeping `"dadjoke" ` in aliases

ItsDrike commented on cogs/fun.py:
```diff
@@ -25,6 +25,74 @@ def __init__(self, bot: Bot) -> None:
+                    )
+                )
+            else:
+                await ctx.send("Lmao! Something went boom! :(")
```
We should include a real error message instead of this, that way we'll at least get the error code and it can be clearly reported, while this message is quite weird for that purpose.

Same goes for lines: 61, 78, 95

janaSunrise committed:
Prepare update fun for merge

janaSunrise committed:
Add missing return annotation

janaSunrise committed:
Solve codefactor errors

ItsDrike reviewed (changes requested):
One more suggestion: You should make a `Game` class (name it however you wish if `Game` isn't sufficient) which will control the game itself and keep this class only for the reaction event and the initial command itself. *(If you have trouble with doing that, you can just ask me on discord)*

Note: These aren't all suggestions I will make, those were only the ones after a quicker look (considering there were 14 suggestions with some really complex ones, that's quite something to say but anyway) I will likely post more suggestions here after the code is cleaned and more readable, hence making it easier to find more mistakes.

janaSunrise commented on bot/cogs/games/ttt.py:
```diff
@@ -0,0 +1,237 @@
+            return "draw"
+        return None
+
+    def tttFindStreaks(self, m: list, xo: str) -> tuple:
+        row = [0, 0, 0]
+        col = [0, 0, 0]
+        dia = [0, 0]
+
+        # Check rows and columns for X streaks
+        for y in range(3):
+            for x in range(3):
+                if m[y][x] == xo:
+                    row[y] += 1
+                    col[x] += 1
+
+        # Check diagonals
+        if m[0][0] == xo:
+            dia[0] += 1
+        if m[1][1] == xo:
+            dia[0] += 1
+            dia[1] += 1
+        if m[2][2] == xo:
+            dia[0] += 1
+        if m[2][0] == xo:
+            dia[1] += 1
+        if m[0][2] == xo:
+            dia[1] += 1
+
+        return (row, col, dia)
```
sure

janaSunrise replied:
but i have a question,
```py
def ttt_ai_think(self, m: list) -> t.Union[tuple, bool]:
        rx, cx, dx = self.ttt_find_streaks(m, "x")
        ro, co, do = self.ttt_find_streaks(m, "o")

        mv = self.ttt_ai_move(2, m, ro, co, do)
        if mv is not False:
            return mv
        mv = self.ttt_ai_move(2, m, rx, cx, dx)
        if mv is not False:
            return mv
        mv = self.ttt_ai_move(1, m, ro, co, do)
        if mv is not False:
            return mv
        return self.ttt_ai_move(1, m, rx, cx, dx)
```
this uses the streak function, what should i do, and when the win function returns a tuple, who should be the winner, how do i find that?

ItsDrike replied:
the readability of this is really bad, also, do we even want AI for tic tac toe here? I can design it in a better way, but if I do it will always either win to draw, it will never lose which may be disappointing for users

ItsDrike committed:
Apply requested changes

ItsDrike committed:
Rework TicTacToe

- Use a custom class for the game
- Use default dict of default dicts which stores `guild` as first key
  and `player` as second key to keep track of the active games. This
  dict is used to store instances of the new Game class
- Use actual 2D list instead of using a flat version
- Implement a better `check_win` function
- Use embed to show the game field and redraw it on each new turn
- Define reactions in a single list (no need to repeat ourselves)
- Temporarely remove the old AI functions, they will be implemented
  later

ItsDrike committed:
Merge branch 'master' into update-fun

ItsDrike committed:
Add CPU logic

ItsDrike committed:
Use emojis for board printing

ItsDrike committed:
Remove redundant else

janaSunrise reviewed (commented):
no issues found :+1:

Faholan commented on bot/cogs/fun.py:
```diff
@@ -33,12 +33,80 @@ def __init__(self, bot: Bot) -> None:
 
+    @command()
+    async def joke(self, ctx: Context) -> None:
+        """Sends a random joke."""
```
```suggestion
        """Send a random joke."""
```

Faholan replied:
Cam is too fast

Faholan commented on bot/cogs/fun.py:
```diff
@@ -43,12 +43,80 @@ def __init__(self, bot: Bot) -> None:
+            if resp.status == 200:
+                data = await resp.json()
+                joke = data["joke"]
+                await ctx.send(
+                    embed=Embed(
+                        description=joke,
+                        color=Color.gold()
+                    )
+                )
```
Yes, this is the recommended behaviour

Faholan commented on bot/cogs/fun.py:
```diff
@@ -43,12 +43,80 @@ def __init__(self, bot: Bot) -> None:
+        """Get a random picture of a duck."""
+        async with self.session.get("https://random-d.uk/api/v2/random") as resp:
+            if resp.status == 200:
+                data = await resp.json()
+                embed = Embed(
+                    title="Random Duck!",
+                    color=Color.gold()
+                )
+                embed.set_image(url=data["url"])
```
Was going to comment this

Faholan commented on bot/cogs/fun.py:
```diff
@@ -33,12 +33,80 @@ def __init__(self, bot: Bot) -> None:
+                data = await resp.json()
+                embed = Embed(
+                    title="Random Duck!",
+                    color=Color.gold()
```
```suggestion
                    color=Color.gold(),
```
Either don't split it, or add a `,` at the end of each line.

Faholan commented on bot/cogs/fun.py:
````diff
@@ -92,10 +160,13 @@ async def image(self, ctx: Context, type: str) -> None:
             )
             embed.set_image(url=nekos.img(type))
             await ctx.send(embed=embed)
+
         except errors.NSFWChannelRequired:
             await ctx.send("Hey dude! Go use this command in a NSFW Channel, this ain't ur home.")
+
         except nekos.errors.InvalidArgument:
             await ctx.send(f"Invalid type! Possible types are : ```{', '.join(config.nsfw_possible)}```")
+
````
Usually, excepts aren't separated by newlines

ItsDrike replied:
I agree, this is janas code, I'll update it.

Faholan commented on bot/cogs/fun.py:
```diff
@@ -178,6 +251,7 @@ def decide_source() -> str:
                     if shibe_get.host == "random.dog":
                         shibe_img = await shibe_get.text()
                         shibe_url = "https://random.dog/" + shibe_img
+
                     elif shibe_get.host == "dog.ceo":
```
elifs shouldn't be separated by a newline

Faholan commented on bot/cogs/fun.py:
```diff
@@ -201,8 +276,9 @@ async def lizard(self, ctx) -> None:
                     liz_em = Embed(colour=0x690E8)
                     liz_em.set_image(url=img["url"])
                     await ctx.send(embed=liz_em)
+
                 else:
```
elses shouldn't be separated by a newline

Faholan commented on bot/cogs/fun.py:
```diff
@@ -257,16 +333,17 @@ async def why(self, ctx: Context) -> None:
-            await ctx.send(embed=e)
+            embed.set_image(url=f"https://robohash.org/{meme}.png")
+            await ctx.send(embed=embed)
+
         except Exception as e:
```
The same comment about except goes for each instance of this

Faholan commented on bot/cogs/games/tic_tac_toe.py:
```diff
@@ -0,0 +1,296 @@
+            self.player = "x"
+
+    async def send_board(self) -> None:
+        """Send embed with the board,"""
```
```suggestion
        """Send the embed with the board."""
```

Faholan commented on bot/cogs/games/tic_tac_toe.py:
```diff
@@ -0,0 +1,296 @@
+        return all("" not in row for row in self.board)
+
+    async def apply_win(self, xo: t.Literal["x", "o"]) -> None:
+        """If win is True, re"""
```
I guess that this docstring isn't complete...

ItsDrike replied:
Lol, how did I not notice that, thanks

ItsDrike committed:
Fix formatting issues in fun.py

Faholan commented on bot/cogs/games/tic_tac_toe.py:
```diff
@@ -0,0 +1,296 @@
+import random
```
Global comment on this file : use `discord.ext.menus` instead of your listener. Your code will be much more readable

ItsDrike replied:
Alright, I'll look into it, note that you can post a global comment  while reviewing like this:
![Peek 2020-07-27 13-41](https://user-images.githubusercontent.com/20902250/88538073-037f9080-d00f-11ea-9d3b-21b3f6be6815.gif)

Faholan replied:
yes

ItsDrike committed:
Update bot/cogs/fun.py

Co-authored-by: Faholan <62927863+Faholan@users.noreply.github.com>

ItsDrike committed:
Update bot/cogs/fun.py

Co-authored-by: Faholan <62927863+Faholan@users.noreply.github.com>

ItsDrike committed:
Update bot/cogs/games/tic_tac_toe.py

Co-authored-by: Faholan <62927863+Faholan@users.noreply.github.com>

ItsDrike committed:
Remove blank lines between excepts

ItsDrike committed:
Remove redundant blank lines

ItsDrike committed:
Merge branch 'update-fun' of https://github.com/The-Codin-Hole/HotWired-Bot into update-fun

ItsDrike committed:
Fix incomplete docstring

ItsDrike committed:
Remove redundant blank line

Faholan committed:
Make Tic Tac Toe a Menu

janaSunrise committed:
Merge branch 'master' into update-fun
